activesupport/lib/active_support/claims.rb

```diff
@@ -0,0 +1,69 @@
+require 'active_support/time'
+
+module ActiveSupport
+  class Claims # :nodoc:
+    class InvalidClaims < StandardError; end
+    class ExpiredClaims < StandardError; end
+    
+    attr_reader :payload, :purpose, :expires_at
+    UNIVERSAL_PURPOSE = 'universal'
+
+    def initialize(payload:, **options)
+      @payload = payload
+      @purpose = self.class.pick_purpose(options)
+      @expires_at = pick_expiration(options)
+    end
+
+    class << self
+      attr_accessor :expires_in
+
+      def verify!(claims, options = {})
```

Comment:
Sure, we can still get at them in the verify methods, but the nice thing about having that stuff contained in `validate` is that anything we add there will automatically improve the verify methods. This means the verify methods don't have to **change**.

To me your changes add too much indirection by putting everything into its own method. What I liked about `validate` was that my eye would scan `verify` and `verify!` (which was easy because they're one line) and see I had to look in `validate`. Looking in `validate` I could see everything I needed at a glance because the helper methods were clearly named - I could even stop reading there.

Without having that `validate` method I find there's too many pieces and the extra method "boundaries" makes it tougher to keep in my head.

`check_purpose?` doesn't capture it for me. What does it mean for a purpose to be checked? How does it pass and how does it fail? `same_purpose?` is clearer IMO.

I also think `unexpired?` is better than `not_expired?`. Though I think it would be better to find a word that focuses on the freshness - you know instead of saying unexpired what's the "positive" way of saying the token is still alive.